#### Main window
'''
    Copyright (C) 2020 Oliver Rodriguez
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
'''
#### @author oliverrdz
#### https://oliverrdz.xyz

from PyQt5 import QtCore, QtGui, QtWidgets
from about import Ui_About

class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(752, 553)
        icon = QtGui.QIcon()
        icon.addPixmap(QtGui.QPixmap("logo.svg"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        MainWindow.setWindowIcon(icon)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.groupBox = QtWidgets.QGroupBox(self.centralwidget)
        self.groupBox.setGeometry(QtCore.QRect(390, 140, 351, 191))
        self.groupBox.setObjectName("groupBox")
        self.formLayoutWidget = QtWidgets.QWidget(self.groupBox)
        self.formLayoutWidget.setGeometry(QtCore.QRect(10, 30, 331, 151))
        self.formLayoutWidget.setObjectName("formLayoutWidget")
        self.formLayout = QtWidgets.QFormLayout(self.formLayoutWidget)
        self.formLayout.setContentsMargins(0, 0, 0, 0)
        self.formLayout.setObjectName("formLayout")
        self.label = QtWidgets.QLabel(self.formLayoutWidget)
        self.label.setObjectName("label")
        self.formLayout.setWidget(0, QtWidgets.QFormLayout.LabelRole, self.label)
        self.box_Eini = QtWidgets.QLineEdit(self.formLayoutWidget)
        self.box_Eini.setReadOnly(False)
        self.box_Eini.setObjectName("box_Eini")
        self.formLayout.setWidget(0, QtWidgets.QFormLayout.FieldRole, self.box_Eini)
        self.box_Efin = QtWidgets.QLineEdit(self.formLayoutWidget)
        self.box_Efin.setObjectName("box_Efin")
        self.formLayout.setWidget(1, QtWidgets.QFormLayout.FieldRole, self.box_Efin)
        self.label_2 = QtWidgets.QLabel(self.formLayoutWidget)
        self.label_2.setObjectName("label_2")
        self.formLayout.setWidget(1, QtWidgets.QFormLayout.LabelRole, self.label_2)
        self.label_3 = QtWidgets.QLabel(self.formLayoutWidget)
        self.label_3.setObjectName("label_3")
        self.formLayout.setWidget(2, QtWidgets.QFormLayout.LabelRole, self.label_3)
        self.label_4 = QtWidgets.QLabel(self.formLayoutWidget)
        self.label_4.setObjectName("label_4")
        self.formLayout.setWidget(3, QtWidgets.QFormLayout.LabelRole, self.label_4)
        self.label_5 = QtWidgets.QLabel(self.formLayoutWidget)
        self.label_5.setObjectName("label_5")
        self.formLayout.setWidget(4, QtWidgets.QFormLayout.LabelRole, self.label_5)
        self.box_sr = QtWidgets.QLineEdit(self.formLayoutWidget)
        self.box_sr.setObjectName("box_sr")
        self.formLayout.setWidget(2, QtWidgets.QFormLayout.FieldRole, self.box_sr)
        self.box_dE = QtWidgets.QLineEdit(self.formLayoutWidget)
        self.box_dE.setObjectName("box_dE")
        self.formLayout.setWidget(3, QtWidgets.QFormLayout.FieldRole, self.box_dE)
        self.box_ns = QtWidgets.QLineEdit(self.formLayoutWidget)
        self.box_ns.setObjectName("box_ns")
        self.formLayout.setWidget(4, QtWidgets.QFormLayout.FieldRole, self.box_ns)
        self.groupBox_2 = QtWidgets.QGroupBox(self.centralwidget)
        self.groupBox_2.setGeometry(QtCore.QRect(10, 10, 351, 321))
        self.groupBox_2.setObjectName("groupBox_2")
        self.formLayoutWidget_2 = QtWidgets.QWidget(self.groupBox_2)
        self.formLayoutWidget_2.setGeometry(QtCore.QRect(10, 30, 331, 310))
        self.formLayoutWidget_2.setObjectName("formLayoutWidget_2")
        self.formLayout_2 = QtWidgets.QFormLayout(self.formLayoutWidget_2)
        self.formLayout_2.setContentsMargins(0, 0, 0, 0)
        self.formLayout_2.setObjectName("formLayout_2")
        self.label_6 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_6.setObjectName("label_6")
        self.formLayout_2.setWidget(0, QtWidgets.QFormLayout.LabelRole, self.label_6)
        self.box_n = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_n.setObjectName("box_n")
        self.formLayout_2.setWidget(0, QtWidgets.QFormLayout.FieldRole, self.box_n)
        self.label_7 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_7.setObjectName("label_7")
        self.formLayout_2.setWidget(1, QtWidgets.QFormLayout.LabelRole, self.label_7)
        self.box_A = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_A.setEnabled(True)
        self.box_A.setObjectName("box_A")
        self.formLayout_2.setWidget(1, QtWidgets.QFormLayout.FieldRole, self.box_A)
        self.label_8 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_8.setObjectName("label_8")
        self.formLayout_2.setWidget(2, QtWidgets.QFormLayout.LabelRole, self.label_8)
        self.box_E0 = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_E0.setObjectName("box_E0")
        self.formLayout_2.setWidget(2, QtWidgets.QFormLayout.FieldRole, self.box_E0)
        self.label_9 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_9.setObjectName("label_9")
        self.formLayout_2.setWidget(3, QtWidgets.QFormLayout.LabelRole, self.label_9)
        self.box_COb = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_COb.setEnabled(False)
        self.box_COb.setObjectName("box_COb")
        self.formLayout_2.setWidget(3, QtWidgets.QFormLayout.FieldRole, self.box_COb)
        self.label_10 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_10.setObjectName("label_10")
        self.formLayout_2.setWidget(4, QtWidgets.QFormLayout.LabelRole, self.label_10)
        self.box_DO = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_DO.setObjectName("box_DO")
        self.formLayout_2.setWidget(4, QtWidgets.QFormLayout.FieldRole, self.box_DO)
        self.label_11 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_11.setObjectName("label_11")
        self.formLayout_2.setWidget(6, QtWidgets.QFormLayout.LabelRole, self.label_11)
        self.box_DR = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_DR.setObjectName("box_DR")
        self.formLayout_2.setWidget(6, QtWidgets.QFormLayout.FieldRole, self.box_DR)
        self.label_13 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_13.setObjectName("label_13")
        self.formLayout_2.setWidget(8, QtWidgets.QFormLayout.LabelRole, self.label_13)
        self.box_ks = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_ks.setEnabled(False)
        self.box_ks.setObjectName("box_ks")
        self.formLayout_2.setWidget(8, QtWidgets.QFormLayout.FieldRole, self.box_ks)
        self.label_14 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_14.setObjectName("label_14")
        self.formLayout_2.setWidget(9, QtWidgets.QFormLayout.LabelRole, self.label_14)
        self.box_alpha = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_alpha.setEnabled(False)
        self.box_alpha.setObjectName("box_alpha")
        self.formLayout_2.setWidget(9, QtWidgets.QFormLayout.FieldRole, self.box_alpha)
        self.label_12 = QtWidgets.QLabel(self.formLayoutWidget_2)
        self.label_12.setObjectName("label_12")
        self.formLayout_2.setWidget(5, QtWidgets.QFormLayout.LabelRole, self.label_12)
        self.box_CRb = QtWidgets.QLineEdit(self.formLayoutWidget_2)
        self.box_CRb.setObjectName("box_CRb")
        self.formLayout_2.setWidget(5, QtWidgets.QFormLayout.FieldRole, self.box_CRb)
        self.groupBox_3 = QtWidgets.QGroupBox(self.centralwidget)
        self.groupBox_3.setGeometry(QtCore.QRect(390, 10, 351, 121))
        self.groupBox_3.setObjectName("groupBox_3")
        self.verticalLayoutWidget = QtWidgets.QWidget(self.groupBox_3)
        self.verticalLayoutWidget.setGeometry(QtCore.QRect(10, 30, 321, 83))
        self.verticalLayoutWidget.setObjectName("verticalLayoutWidget")
        self.verticalLayout = QtWidgets.QVBoxLayout(self.verticalLayoutWidget)
        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout.setObjectName("verticalLayout")
        self.Nernst = QtWidgets.QRadioButton(self.verticalLayoutWidget)
        self.Nernst.setChecked(True)
        self.Nernst.setObjectName("Nernst")
        self.verticalLayout.addWidget(self.Nernst)
        self.Irrev = QtWidgets.QRadioButton(self.verticalLayoutWidget)
        self.Irrev.setObjectName("Irrev")
        self.verticalLayout.addWidget(self.Irrev)
        self.BV = QtWidgets.QRadioButton(self.verticalLayoutWidget)
        self.BV.setObjectName("BV")
        self.verticalLayout.addWidget(self.BV)
        self.horizontalLayoutWidget = QtWidgets.QWidget(self.centralwidget)
        self.horizontalLayoutWidget.setGeometry(QtCore.QRect(390, 360, 351, 31))
        self.horizontalLayoutWidget.setObjectName("horizontalLayoutWidget")
        self.horizontalLayout = QtWidgets.QHBoxLayout(self.horizontalLayoutWidget)
        self.horizontalLayout.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.btnSimulate = QtWidgets.QPushButton(self.horizontalLayoutWidget)
        self.btnSimulate.setObjectName("btnSimulate")
        self.horizontalLayout.addWidget(self.btnSimulate)
        self.groupBox_4 = QtWidgets.QGroupBox(self.centralwidget)
        self.groupBox_4.setGeometry(QtCore.QRect(10, 340, 351, 101))
        self.groupBox_4.setObjectName("groupBox_4")
        self.formLayoutWidget_3 = QtWidgets.QWidget(self.groupBox_4)
        self.formLayoutWidget_3.setGeometry(QtCore.QRect(10, 30, 331, 61))
        self.formLayoutWidget_3.setObjectName("formLayoutWidget_3")
        self.formLayout_3 = QtWidgets.QFormLayout(self.formLayoutWidget_3)
        self.formLayout_3.setContentsMargins(0, 0, 0, 0)
        self.formLayout_3.setObjectName("formLayout_3")
        self.label_15 = QtWidgets.QLabel(self.formLayoutWidget_3)
        self.label_15.setObjectName("label_15")
        self.formLayout_3.setWidget(0, QtWidgets.QFormLayout.LabelRole, self.label_15)
        self.box_Cd = QtWidgets.QLineEdit(self.formLayoutWidget_3)
        self.box_Cd.setEnabled(False)
        self.box_Cd.setObjectName("box_Cd")
        self.formLayout_3.setWidget(0, QtWidgets.QFormLayout.FieldRole, self.box_Cd)
        self.label_16 = QtWidgets.QLabel(self.formLayoutWidget_3)
        self.label_16.setObjectName("label_16")
        self.formLayout_3.setWidget(1, QtWidgets.QFormLayout.LabelRole, self.label_16)
        self.box_Ru = QtWidgets.QLineEdit(self.formLayoutWidget_3)
        self.box_Ru.setEnabled(False)
        self.box_Ru.setObjectName("box_Ru")
        self.formLayout_3.setWidget(1, QtWidgets.QFormLayout.FieldRole, self.box_Ru)
        self.verticalLayoutWidget_2 = QtWidgets.QWidget(self.centralwidget)
        self.verticalLayoutWidget_2.setGeometry(QtCore.QRect(390, 410, 351, 31))
        self.verticalLayoutWidget_2.setObjectName("verticalLayoutWidget_2")
        self.verticalLayout_2 = QtWidgets.QVBoxLayout(self.verticalLayoutWidget_2)
        self.verticalLayout_2.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout_2.setObjectName("verticalLayout_2")
        self.btnPlot = QtWidgets.QPushButton(self.verticalLayoutWidget_2)
        self.btnPlot.setEnabled(False)
        self.btnPlot.setObjectName("btnPlot")
        self.verticalLayout_2.addWidget(self.btnPlot)
        self.verticalLayoutWidget_3 = QtWidgets.QWidget(self.centralwidget)
        self.verticalLayoutWidget_3.setGeometry(QtCore.QRect(9, 460, 731, 41))
        self.verticalLayoutWidget_3.setObjectName("verticalLayoutWidget_3")
        self.verticalLayout_3 = QtWidgets.QVBoxLayout(self.verticalLayoutWidget_3)
        self.verticalLayout_3.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout_3.setObjectName("verticalLayout_3")
        self.progressBar = QtWidgets.QProgressBar(self.verticalLayoutWidget_3)
        self.progressBar.setEnabled(False)
        self.progressBar.setProperty("value", 0)
        self.progressBar.setObjectName("progressBar")
        self.verticalLayout_3.addWidget(self.progressBar)
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 752, 22))
        self.menubar.setObjectName("menubar")
        self.menuNew = QtWidgets.QMenu(self.menubar)
        self.menuNew.setObjectName("menuNew")
        self.menuHelp = QtWidgets.QMenu(self.menubar)
        self.menuHelp.setObjectName("menuHelp")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)
        self.actionCV = QtWidgets.QAction(MainWindow)
        self.actionCV.setObjectName("actionCV")
        self.actionChronoamperometry = QtWidgets.QAction(MainWindow)
        self.actionChronoamperometry.setEnabled(False)
        self.actionChronoamperometry.setObjectName("actionChronoamperometry")
        self.actionSampled_current_voltammetry = QtWidgets.QAction(MainWindow)
        self.actionSampled_current_voltammetry.setEnabled(False)
        self.actionSampled_current_voltammetry.setObjectName("actionSampled_current_voltammetry")
        self.actionAbout = QtWidgets.QAction(MainWindow)
        self.actionAbout.setObjectName("actionAbout")
        self.menuNew.addAction(self.actionCV)
        self.menuNew.addAction(self.actionChronoamperometry)
        self.menuNew.addAction(self.actionSampled_current_voltammetry)
        self.menuHelp.addAction(self.actionAbout)
        self.menubar.addAction(self.menuNew.menuAction())
        self.menubar.addAction(self.menuHelp.menuAction())

        # Connect all buttons:
        self.btnSimulate.clicked.connect(self.clickSimulate)
        self.btnPlot.clicked.connect(self.clickPlot)
        
        # Connect boundary conditions (radio buttons):
        self.Nernst.toggled.connect(self.radioClicked)
        self.Irrev.toggled.connect(self.radioClicked)
        self.BV.toggled.connect(self.radioClicked)
        
        # Connect menu:
        self.actionCV.triggered.connect(lambda: self.resetCV())
        self.actionAbout.triggered.connect(lambda: self.open_About())

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "SP"))
        self.groupBox.setTitle(_translate("MainWindow", "Cyclic voltammetry parameters:"))
        self.label.setText(_translate("MainWindow", "Initial potential (V):"))
        self.box_Eini.setText(_translate("MainWindow", "-0.5"))
        self.box_Efin.setText(_translate("MainWindow", "0.5"))
        self.label_2.setText(_translate("MainWindow", "Final potential (V):"))
        self.label_3.setText(_translate("MainWindow", "Scan rate (V/s):"))
        self.label_4.setText(_translate("MainWindow", "Potential increment (V):"))
        self.label_5.setText(_translate("MainWindow", "Number of sweeps:"))
        self.box_sr.setText(_translate("MainWindow", "1"))
        self.box_dE.setText(_translate("MainWindow", "0.01"))
        self.box_ns.setText(_translate("MainWindow", "2"))
        self.groupBox_2.setTitle(_translate("MainWindow", "Electrochemistry parameters:"))
        self.label_6.setText(_translate("MainWindow", "Number of electrons:"))
        self.box_n.setText(_translate("MainWindow", "1"))
        self.label_7.setText(_translate("MainWindow", "<html><head/><body><p>Geometrical area (cm<span style=\" vertical-align:super;\">2</span>):</p></body></html>"))
        self.box_A.setText(_translate("MainWindow", "1"))
        self.label_8.setText(_translate("MainWindow", "Standard potential (V):"))
        self.box_E0.setText(_translate("MainWindow", "0"))
        self.label_9.setText(_translate("MainWindow", "<html><head/><body><p>Bulk concentration of O (mol/cm<span style=\" vertical-align:super;\">3</span>):</p></body></html>"))
        self.box_COb.setText(_translate("MainWindow", "0"))
        self.label_10.setText(_translate("MainWindow", "<html><head/><body><p>Diffusion coefficient of O (cm<span style=\" vertical-align:super;\">2</span>/s):</p></body></html>"))
        self.box_DO.setText(_translate("MainWindow", "1e-5"))
        self.label_11.setText(_translate("MainWindow", "<html><head/><body><p>Diffusion coefficient of R (cm<span style=\" vertical-align:super;\">2</span>/s):</p></body></html>"))
        self.box_DR.setText(_translate("MainWindow", "1e-5"))
        self.label_13.setText(_translate("MainWindow", "Standard rate constant (cm/s):"))
        self.box_ks.setText(_translate("MainWindow", "1e-3"))
        self.label_14.setText(_translate("MainWindow", "Transfer coefficient:"))
        self.box_alpha.setText(_translate("MainWindow", "0.5"))
        self.label_12.setText(_translate("MainWindow", "<html><head/><body><p>Bulk concentration of R (mol/cm<span style=\" vertical-align:super;\">3</span>):</p></body></html>"))
        self.box_CRb.setText(_translate("MainWindow", "5e-6"))
        self.groupBox_3.setTitle(_translate("MainWindow", "Surface concentration:"))
        self.Nernst.setText(_translate("MainWindow", "Nernstian"))
        self.Irrev.setText(_translate("MainWindow", "Butler-Volmer irreversible oxidation"))
        self.BV.setText(_translate("MainWindow", "Butler-Volmer quasi reversible"))
        self.btnSimulate.setText(_translate("MainWindow", "Simulate"))
        self.groupBox_4.setTitle(_translate("MainWindow", "Non-faradaic parameters:"))
        self.label_15.setText(_translate("MainWindow", "Double layer capacitance (F):"))
        self.box_Cd.setText(_translate("MainWindow", "0"))
        self.label_16.setText(_translate("MainWindow", "<html><head/><body><p>Solution resistance (ohms):</p></body></html>"))
        self.box_Ru.setText(_translate("MainWindow", "0"))
        self.btnPlot.setText(_translate("MainWindow", "Plot"))
        self.menuNew.setTitle(_translate("MainWindow", "New"))
        self.menuHelp.setTitle(_translate("MainWindow", "Help"))
        self.actionCV.setText(_translate("MainWindow", "Cyclic Voltammetry"))
        self.actionChronoamperometry.setText(_translate("MainWindow", "Chronoamperometry"))
        self.actionSampled_current_voltammetry.setText(_translate("MainWindow", "Sampled current voltammetry"))
        self.actionAbout.setText(_translate("MainWindow", "About"))

    def open_About(self):
        dialog = QtWidgets.QDialog()
        dialog.ui = Ui_About()
        dialog.ui.setupUi(dialog)
        dialog.exec_()
        dialog.show()
